### PR TITLE
[Serializer][XmlEncoder] Allow removing empty tags in generated XML

### PR DESCRIPTION
**Pull request by amoiraud**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #20398 
| License       | MIT
| Doc PR        | ~

Allow a new option in $context of XmlEncoder.php to remove empty tags if $context['remove_empty_tags'] setted to true, changing this : 

```xml
    <node>
         <subnode>Value</subnode>
         <emptysubnode/>
    </node>
```

To this :

```xml
    <node>
         <subnode>Value</subnode>
    </node>
```

### CODE REVIEW

File: src/Symfony/Component/Serializer/Encoder/XmlEncoder.php [hunk 1]
```diff
@@ -393,7 +393,9 @@ private function buildXml(\DOMNode $parentNode, $data, $xmlRootNodeName = null)
                 } elseif (is_numeric($key) || !$this->isElementNameValid($key)) {
                     $append = $this->appendNode($parentNode, $data, 'item', $key);
                 } else {
-                    $append = $this->appendNode($parentNode, $data, $key);
+                    if ($data !== null || !isset($this->context['remove_empty_tags']) || $this->context['remove_empty_tags'] === false) {
```
Comment (xabbuh) [inline src/Symfony/Component/Serializer/Encoder/XmlEncoder.php]: should be `elseif`

Comment (xabbuh) [inline src/Symfony/Component/Serializer/Encoder/XmlEncoder.php]: and please also use a Yoda condition here: `null !== $data`

**Review by xabbuh — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Serializer/Encoder/XmlEncoder.php (hunk 1 above)
Comment (amoiraud) [inline src/Symfony/Component/Serializer/Encoder/XmlEncoder.php]: It's done :

```php
} elseif (null !== $data || !isset($this->context['remove_empty_tags']) || false === $this->context['remove_empty_tags']) {
    $append = $this->appendNode($parentNode, $data, $key);
}
```

**Review by xabbuh — APPROVED** [review]

File: src/Symfony/Component/Serializer/Tests/Encoder/XmlEncoderTest.php [hunk 2]
```diff
@@ -138,6 +138,30 @@ public function testEncodeXmlAttributes()
...
     }
 
+    public function testEncodeRemovingEmptyTags()
+    {
+        $array = array('person' => array('firstname' => 'Peter', 'lastname' => null));
+
+        $expected = '<?xml version="1.0"?>'."\n".
+            '<response><person><firstname>Peter</firstname></person></response>'."\n";
+
+        $context = array(
```
Comment (dunglas) [inline src/Symfony/Component/Serializer/Tests/Encoder/XmlEncoderTest.php]: Can be inline

**Comment by dunglas** [comment]
👍

**Comment by dunglas** [comment]
Can you open a doc PR please @amoiraud

**Comment by dunglas** [comment]
Thank you @amoiraud.

**Comment by dunglas** [comment]
@amoiraud, I suggest to add a new section about the XmlEncoder in https://github.com/symfony/symfony-docs/blob/master/components/serializer.rst and list its available options.

**Pull request closed without merging**